crates/q_cli/src/cli/chat/tools/mod.rs

```diff
@@ -0,0 +1,169 @@
+                });
+            },
+        })
+    }
+
+    /// The display name of a tool
+    pub fn display_name(&self) -> String {
+        match self {
+            ToolE::FsRead(_) => FsRead::display_name(),
+            ToolE::FsWrite(_) => FsWrite::display_name(),
+            ToolE::ExecuteBash(_) => ExecuteBash::display_name(),
+            ToolE::UseAws(_) => UseAws::display_name(),
+        }
+    }
+
+    /// Invokes the tool asynchronously
+    pub async fn invoke(&self, context: &Context, updates: &mut impl Write) -> Result<InvokeOutput> {
+        match self {
+            ToolE::FsRead(fs_read) => fs_read.invoke(context, updates).await,
+            ToolE::FsWrite(fs_write) => fs_write.invoke(context, updates).await,
+            ToolE::ExecuteBash(execute_bash) => execute_bash.invoke(updates).await,
+            ToolE::UseAws(use_aws) => use_aws.invoke(context, updates).await,
+        }
+    }
+
+    /// Queues up a tool's intention in a human readable format
+    pub fn show_readable_intention(&self, updates: &mut impl Write) -> Result<()> {
+        match self {
+            ToolE::FsRead(fs_read) => fs_read.show_readable_intention(updates),
+            ToolE::FsWrite(fs_write) => fs_write.show_readable_intention(updates),
+            ToolE::ExecuteBash(execute_bash) => execute_bash.show_readable_intention(updates),
+            ToolE::UseAws(use_aws) => use_aws.show_readable_intention(updates),
+        }
+    }
+
+    /// Validates the tool with the arguments supplied
+    pub async fn validate(&mut self, ctx: &Context) -> Result<()> {
+        match self {
+            ToolE::FsRead(fs_read) => fs_read.validate(ctx).await,
+            ToolE::FsWrite(fs_write) => fs_write.validate(ctx).await,
+            ToolE::ExecuteBash(execute_bash) => execute_bash.validate(ctx).await,
+            ToolE::UseAws(use_aws) => use_aws.validate(ctx).await,
+        }
+    }
+}
+
+/// A tool specification to be sent to the model as part of a conversation. Maps to
+/// [BedrockToolSpecification].
+#[derive(Debug, Clone, Deserialize)]
+pub struct ToolSpec {
+    pub name: String,
+    pub description: String,
+    pub input_schema: InputSchema,
+}
+
+/// The schema specification describing a tool's fields.
+#[derive(Debug, Clone, Deserialize)]
+pub struct InputSchema(pub serde_json::Value);
+
+/// The output received from invoking a [Tool].
+#[derive(Debug, Default)]
+pub struct InvokeOutput {
+    pub output: OutputKind,
+}
+
+#[non_exhaustive]
+#[derive(Debug)]
+pub enum OutputKind {
+    Text(String),
+    Json(serde_json::Value),
+}
+
+impl Default for OutputKind {
+    fn default() -> Self {
+        Self::Text(String::new())
+    }
+}
+
+pub fn serde_value_to_document(value: serde_json::Value) -> Document {
+    match value {
+        serde_json::Value::Null => Document::Null,
+        serde_json::Value::Bool(bool) => Document::Bool(bool),
+        serde_json::Value::Number(number) => {
+            if number.is_f64() {
```

Comment:
I would use the as methods and now unwrap: https://docs.rs/serde_json/latest/serde_json/struct.Number.html#method.as_i64